### PR TITLE
V22F-724 cover mediator

### PR DESCRIPTION
**Pull request by KungFuDonkey**

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 1]
```diff
@@ -3,21 +3,28 @@
...
 #include <sys/types.h>
 #include <sys/stat.h>
 #define _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING 1
 #include <experimental/filesystem>
 #include "../rppUtils/RppUtils.hpp"
 
-#define MockMediator Mediator<DecisionMakerMock>
+#define MockMediator Mediator<SDecisionMakerMock>
+
+#define MockMediator2 Mediator<DecisionMakerMock<SDAConfig>>
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp]: Add comments to explain what the difference between MockMediator and MockMediator2 is

Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp]: Clarify names or rename.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 2]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+    ASSERT_EQ(arr.ControlSteering, pControl.ControlSteering);
+    ASSERT_EQ(arr.ControlGas, pControl.ControlGas);
+}
+
+/// @brief Tests the Mediator ParticipantControlSettings for every possible boolean combination (first 3)
+BEGIN_TEST_COMBINATORIAL(MediatorTests, PControlSettings1)
+bool booleans[] = {false, true};
+END_TEST_COMBINATORIAL3(PControlTest1Mediator, booleans, 2, booleans, 2, booleans, 2)
+
+/// @brief         Tests if the v sets and gets the other pControl settings correctly
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:155]: What is "the v"? What are "the other pControl settings"?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 3]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+/// @brief Tests the Mediator ParticipantControlSettings for every possible boolean combination (last 3)
+BEGIN_TEST_COMBINATORIAL(MediatorTests, PControlSettings2)
+bool booleans[] = {false, true};
+END_TEST_COMBINATORIAL3(PControlTest2Mediator, booleans, 2, booleans, 2, booleans, 2)
+
+/// @brief Tests if the Mediator sets and gets the MaxTime correctly
+TEST(MediatorTests, MaxTimeTest)
+{
+    Random random;
+    for (int i = 0; i < 20; i++)
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:181]: Make 20 a define for `NUMBER_OF_TESTS`

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 4]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+        ASSERT_EQ(maxTime, config.GetMaxTime());
+    }
+}
+
+/// @brief Tests if the Mediator sets and gets the UserID correctly
+TEST(MediatorTests, UserIDTest)
+{
+    Random random;
+    char buf[32];
+    for (int i = 0; i < 20; i++)
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:197]: Same as above, define number of tests.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 5]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+            path[i] = c;
+        }
+        path[length - 1] = '\0';
+        SDAConfigMediator::GetInstance()->SetBlackBoxFilePath(path);
+        const SDAConfig config = SDAConfigMediator::GetInstance()->GetDecisionMaker()->Config;
+        const char* configPath = config.GetBlackBoxFilePath();
+        for (int i = 0; i < length; i++)
+        {
+            ASSERT_EQ(path[i], configPath[i]);
+        }
```
Comment (larsdekwant) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:235]: a while loop in a for loop in a for loop is not according to the code standards. Extract into functions.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 6]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+    ASSERT_EQ(arr.Text, indicator.Text);
+}
+
+/// @brief Tests the Mediator IndicatorSetting for every possible boolean combination
+BEGIN_TEST_COMBINATORIAL(MediatorTests, IndicatorSettings)
+bool booleans[] = {false, true};
+END_TEST_COMBINATORIAL3(IndicatorTestMediator, booleans, 2, booleans, 2, booleans, 2)
+
+/// @brief         Tests if the Mediator sets and gets the participant control settings correctly
+/// @param p_bool1 First  bool
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:134]: Use descriptive variable names and documentation comments.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 7]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+    ASSERT_EQ(arr.ControlGas, pControl.ControlGas);
+}
+
+/// @brief Tests the Mediator ParticipantControlSettings for every possible boolean combination (first 3)
+BEGIN_TEST_COMBINATORIAL(MediatorTests, PControlSettings1)
+bool booleans[] = {false, true};
+END_TEST_COMBINATORIAL3(PControlTest1Mediator, booleans, 2, booleans, 2, booleans, 2)
+
+/// @brief         Tests if the v sets and gets the other pControl settings correctly
+/// @param p_bool1 First  bool
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:156]: Use descriptive variable names and documentation comments.

**Review by larsdekwant — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp (hunk 3 above)
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:181]: Why is this test performed 20 times? Please add a comment explaining why.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp (hunk 4 above)
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:197]: Why is this test performed 20 times? Please add a comment explaining why.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 8]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+        const SDAConfig config = SDAConfigMediator::GetInstance()->GetDecisionMaker()->Config;
+        ASSERT_EQ(buf, config.GetUserId());
+    }
+}
+
+/// @brief Tests if the Mediator sets and gets the BlackBoxFilePath correctly
+TEST(MediatorTests, BlackBoxFilePathTest)
+{
+    Random random;
+    for (int j = 0; j <= 10; j++)
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:213]: Why is this test performed 10 times? Please add a comment explaining why.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 9]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+        const char* configPath = config.GetBlackBoxFilePath();
+        for (int i = 0; i < length; i++)
+        {
+            ASSERT_EQ(path[i], configPath[i]);
+        }
+    }
+}
+
+/// @brief         Tests if the Mediator sets and gets the DataCollectionSetting correctly
+/// @param p_bool1 First bool
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:240]: Use descriptive variable names and documentation comments.

**Review by Sindarin27 — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 10]
```diff
@@ -29,43 +38,226 @@ TEST(MediatorTests, GetDistributedMediatorTemplated)
...
+            } while (c == '\0');
+            path[i] = c;
+        }
+        path[length - 1] = '\0';
+        SDAConfigMediator::GetInstance()->SetBlackBoxFilePath(path);
+        const SDAConfig config = SDAConfigMediator::GetInstance()->GetDecisionMaker()->Config;
+        const char* configPath = config.GetBlackBoxFilePath();
+        for (int i = 0; i < length; i++)
+        {
+            ASSERT_EQ(path[i], configPath[i]);
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:234]: Can ASSERT_STREQ be used instead? If not, ensure to check the length of both strings as well: if configPath is longer than length, this is not caught by the test.

**Review by Sindarin27 — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp [hunk 11]
```diff
@@ -3,21 +3,32 @@
...
 #include "../rppUtils/RppUtils.hpp"
 
-#define MockMediator Mediator<DecisionMakerMock>
+/// @brief A mediator that uses the standard SDecisionMakerMock
+#define MockMediator Mediator<SDecisionMakerMock>
+
+/// @brief A mediator that uses SDAConfig in DecisionmakerMock internally
+#define SDAConfigMediator Mediator<DecisionMakerMock<SDAConfig>>
+
+#define TEST_AMOUNT 20
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/MediatorTests.cpp:20]: Add a comment explaining the defintion of this define, e.g. "The number of trials to run for nondeterministic tests"

**Review by Sindarin27 — CHANGES_REQUESTED** [review]

**Pull request merged**